jitbuilder/lljb/src/IRVisitor.cpp

```diff
@@ -0,0 +1,559 @@
+   _methodBuilder->defineFunction(callee,numParams, paramTypes.data());
+   result = _builder->Call(calleeName, numParams, params.data());
+   _methodBuilder->mapIRtoIlValue(&I, result);
+   }
+
+void
+IRVisitor::visitCastInst(llvm::CastInst &I)
+   {
+   TR::IlValue * srcVal = getIlValue(I.getOperand(0));
+   TR::IlType * toIlType = _methodBuilder->getIlType(_td,I.getDestTy());
+   TR::IlValue * result = _builder->ConvertTo(toIlType, srcVal);
+   _methodBuilder->mapIRtoIlValue(&I, result);
+   }
+
+void
+IRVisitor::visitZExtInst(llvm::ZExtInst &I)
+   {
+   TR::IlValue * srcVal = getIlValue(I.getOperand(0));
+   TR::IlType * toIlType = _methodBuilder->getIlType(_td,I.getDestTy());
+   TR::IlValue * result = _builder->UnsignedConvertTo(toIlType, srcVal);
+   _methodBuilder->mapIRtoIlValue(&I, result);
+   }
+
+void
+IRVisitor::visitGetElementPtrInst(llvm::GetElementPtrInst &I)
+   {
+   TR::IlValue * ilValue = nullptr;
+   if (I.getSourceElementType()->isStructTy())
+      {
+      llvm::ConstantInt * indextConstantInt = llvm::dyn_cast<llvm::ConstantInt>(I.getOperand(2));
+      unsigned elementIndex = indextConstantInt->getZExtValue();
+      ilValue = _builder->StructFieldInstanceAddress(I.getSourceElementType()->getStructName().data(),
+                                                               _methodBuilder->getMemberNameFromIndex(elementIndex),
+                                                               getIlValue(I.getOperand(0)));
+      }
+   else if (I.getSourceElementType()->isArrayTy())
+      {
+      if (I.getSourceElementType()->getArrayElementType()->isArrayTy())
+         {
+         llvm::ConstantInt * indextConstantInt = llvm::dyn_cast<llvm::ConstantInt>(I.getOperand(2));
+         int64_t elementIndex = indextConstantInt->getSExtValue();
+         elementIndex *= I.getSourceElementType()->getArrayNumElements();
+         ilValue =
+               _builder->IndexAt(
+                  _td->PointerTo(
+                     _methodBuilder->getIlType(_td,I.getSourceElementType()->getArrayElementType())),
+                  getIlValue(I.getOperand(0)),
+                  _builder->ConstInt32(elementIndex));
+         }
+      else
+         {
+         TR::IlValue * elementIndex = getIlValue(I.getOperand(2));
+         ilValue =
+               _builder->IndexAt(
+                  _td->PointerTo(
+                     _methodBuilder->getIlType(_td,I.getSourceElementType()->getArrayElementType())),
+                  getIlValue(I.getOperand(0)),
+                  elementIndex);
+         }
+      }
+   else
+      {
+      assert((I.getNumOperands() == 2) && "unhandled getElementPtr case");
+      ilValue = _builder->IndexAt(
+         _td->PointerTo(
+               _methodBuilder->getIlType(_td, I.getSourceElementType())),
+         getIlValue(I.getOperand(0)),
+         getIlValue(I.getOperand(1)));
+      }
+   _methodBuilder->mapIRtoIlValue(&I, ilValue);
+   }
+
+void
+IRVisitor::visitPHINode(llvm::PHINode &I)
```

Comment:
As a general comment (irony coming), I think you could improve on the number of comments in this file.

But I think this function in particular deserves some comments. I understand what it's doing, but not sure I understand why (probably due to lack of llvm familiarity).

Reply:
I agree. I have added *some* comments... hope that helps :)